src/imitation/data/types.py

```diff
@@ -21,7 +26,167 @@
 
 T = TypeVar("T")
 
+TensorVar = TypeVar("TensorVar", np.ndarray, th.Tensor)
+
+
 AnyPath = Union[str, bytes, os.PathLike]
+
+
+@dataclasses.dataclass(frozen=True)
+class DictObs:
```

Comment:
This is the most important choice in the PR: having a separate class to store observations in a dictionary space, instead of just keeping them in a `dict[str, np.ndarray]` format. I'm honestly not sure if it's the correct call.

The main advantage is that it allows code like `len(obs)` to mean a consistent thing regardless of if the observations are in a dictionary space or not. The alternative would probably be to have a collection of utilities that perform a similar role, such as a `obs_len` function that checks if `obs` is a dictionary.

Imo the main disadvantage is complexity bloat of having an additional class. It's sometimes also occasionally to shuffle observations in and out of DictObs objects.